src/managed/Microsoft.Extensions.DependencyModel/JsonTextReaderExtensions.Netstandard2_0.cs

```diff
@@ -0,0 +1,139 @@
+        {
+            reader.Read();
+            CheckStartObject(ref reader);
+        }
+
+        internal static void CheckStartObject(ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                throw CreateUnexpectedException(ref reader, "{");
+            }
+        }
+
+        internal static void CheckEndObject(this ref Utf8JsonReader reader)
+        {
+            if (reader.TokenType != JsonTokenType.EndObject)
+            {
+                throw CreateUnexpectedException(ref reader, "}");
+            }
+        }
+
+        internal static string[] ReadStringArray(this ref Utf8JsonReader reader)
+        {
+            reader.Read();
+            if (reader.TokenType != JsonTokenType.StartArray)
+            {
+                throw CreateUnexpectedException(ref reader,"[");
+            }
+
+            var items = new List<string>();
+
+            while (reader.Read() && reader.TokenType == JsonTokenType.String)
+            {
+                items.Add(reader.GetStringValue());
+            }
+
+            if (reader.TokenType != JsonTokenType.EndArray)
+            {
+                throw CreateUnexpectedException(ref reader, "]");
+            }
+
+            return items.ToArray();
+        }
+
+        internal static string ReadAsString(this ref Utf8JsonReader reader)
+        {
+            reader.Read();
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                throw CreateUnexpectedException(ref reader, "a JSON string token");
+            }
+            return reader.GetStringValue();
+        }
+
+        internal static bool ReadAsBoolean(this ref Utf8JsonReader reader)
+        {
+            reader.Read();
+            if (reader.TokenType != JsonTokenType.True && reader.TokenType != JsonTokenType.False)
+            {
+                throw CreateUnexpectedException(ref reader, "a JSON true or false literal token");
+            }
+            return reader.GetBooleanValue();
+        }
+
+        internal static bool ReadAsNullableBoolean(this ref Utf8JsonReader reader, bool defaultValue)
+        {
+            reader.Read();
+            if (reader.TokenType != JsonTokenType.True && reader.TokenType != JsonTokenType.False)
+            {
+                if (reader.TokenType == JsonTokenType.Null)
+                {
+                    return defaultValue;
+                }
+                throw CreateUnexpectedException(ref reader, "a JSON true or false literal token");
+            }
+            return reader.GetBooleanValue();
+        }
+
+        internal static Exception CreateUnexpectedException(ref Utf8JsonReader reader, string expected)
+        {
+            return new FormatException($"Unexpected character encountered, excepted '{expected}' " +
+                                       $"at line {reader.CurrentState._lineNumber} position {reader.CurrentState._bytePositionInLine}");
```

Comment:
The `Utf8JsonReader` doesn't track/create a path.

Reply:
> IMO you should only be using public APIs from the reader.

Agreed.

However, I don't really understand the need for this method. Why not rely on the exceptions thrown by the JsonReader itself? But, you are right that we may want to expose these publicly.

cc @eerhardt

Reply:
Looks like the code was that way since it was migrated from using `JObject` to simply the `JsonTextReader`.

https://github.com/dotnet/cli/commit/a02d002352517994a53b5e232c57db1c9724599a

I assume it is so we can give "good" exception messages about exactly where we expect a new object/array/etc. What is the exception message you get when you rely on the exceptions thrown by the JsonReader itself?

Reply:
> If position info isn't already public then this is a good argument for providing it. If you need it here then other users of the reader will also likely use it.

https://github.com/dotnet/corefx/issues/34768